File: 19. Circular deque using linked list.cpp
/*
 Task: Implement circular deque using doubly linked list
*/

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int val;
    Node* prev;
    Node* next;
    Node(int v) : val(v), prev(nullptr), next(nullptr) {}
};

class CircularDeque
{
    Node* head;
    Node* tail;
    int cap;
    int sz;

   public:
    CircularDeque(int k)
    {
        head = tail = nullptr;
        cap = k;
        sz = 0;
    }

    bool push_front(int x)
    {
        if (sz == cap)
            return false;
        Node* n = new Node(x);
        if (!head)
        {
            head = tail = n;
        }
        else
        {
            n->next = head;
            head->prev = n;
            head = n;
        }
        sz++;
        return true;
    }

    bool push_back(int x)
    {
        if (sz == cap)
            return false;
        Node* n = new Node(x);
        if (!tail)
        {
            head = tail = n;
        }
        else
        {
            tail->next = n;
            n->prev = tail;
            tail = n;
        }
        sz++;
        return true;
    }

    bool pop_front()
    {
        if (sz == 0)
            return false;
        Node* t = head;
        head = head->next;
        if (head)
            head->prev = nullptr;
        else
            tail = nullptr;
        delete t;
        sz--;
        return true;
    }

    bool pop_back()
    {
        if (sz == 0)
            return false;
        Node* t = tail;
        tail = tail->prev;
        if (tail)
            tail->next = nullptr;
        else
            head = nullptr;
        delete t;
        sz--;
        return true;
    }

    int front()
    {
        if (sz == 0)
            throw runtime_error("empty");
        return head->val;
    }

    int back()
    {
        if (sz == 0)
            throw runtime_error("empty");
        return tail->val;
    }

    int size() { return sz; }
};

int main()
{
    CircularDeque dq(3);
    dq.push_back(1);
    dq.push_back(2);
    dq.push_front(0);
    cout << dq.front() << " " << dq.back() << "\n";
    dq.pop_back();
    cout << dq.back() << "\n";
    return 0;
}

// Time Complexity: O(1) per operation
// Space Complexity: O(k)

File: 19. Circular deque using linked list.py
"""
Task: Circular deque using doubly linked list (duplicate of C++ example)
Time Complexity: O(1)
Space Complexity: O(k)
"""

class Node:
    def __init__(self,val):
        self.val=val
        self.prev=None
        self.next=None

class CircularDeque:
    def __init__(self,k):
        self.head=None; self.tail=None; self.cap=k; self.sz=0

    def push_front(self,x):
        if self.sz==self.cap: return False
        n=Node(x)
        if not self.head:
            self.head=self.tail=n
        else:
            n.next=self.head; self.head.prev=n; self.head=n
        self.sz+=1; return True

    def push_back(self,x):
        if self.sz==self.cap: return False
        n=Node(x)
        if not self.tail:
            self.head=self.tail=n
        else:
            self.tail.next=n; n.prev=self.tail; self.tail=n
        self.sz+=1; return True

    def pop_front(self):
        if self.sz==0: return False
        t=self.head; self.head=self.head.next
        if self.head: self.head.prev=None
        else: self.tail=None
        self.sz-=1; return True

    def pop_back(self):
        if self.sz==0: return False
        t=self.tail; self.tail=self.tail.prev
        if self.tail: self.tail.next=None
        else: self.head=None
        self.sz-=1; return True

if __name__=="__main__":
    dq=CircularDeque(3); dq.push_back(1); dq.push_back(2); dq.push_front(0); print('ok')

# Task: Circular deque using linked list
